ixmp/backend/jdbc.py

```diff
@@ -0,0 +1,649 @@
+from collections import ChainMap
+from collections.abc import Collection, Iterable
+from functools import lru_cache
+import os
+from pathlib import Path
+import re
+from types import SimpleNamespace
+
+import jpype
+from jpype import JClass
+import numpy as np
+import pandas as pd
+
+from ixmp.config import _config
+from ixmp.utils import islistable, logger
+from ixmp.backend.base import Backend, FIELDS
+
+
+# Map of Python to Java log levels
+LOG_LEVELS = {
+    'CRITICAL': 'ALL',
+    'ERROR': 'ERROR',
+    'WARNING': 'WARN',
+    'INFO': 'INFO',
+    'DEBUG': 'DEBUG',
+    'NOTSET': 'OFF',
+}
+
+# Java classes, loaded by start_jvm(). These become available as e.g.
+# java.IxException or java.HashMap.
+java = SimpleNamespace()
+
+JAVA_CLASSES = [
+    'at.ac.iiasa.ixmp.exceptions.IxException',
+    'at.ac.iiasa.ixmp.objects.Scenario',
+    'at.ac.iiasa.ixmp.objects.TimeSeries.TimeSpan',
+    'at.ac.iiasa.ixmp.Platform',
+    'java.lang.Double',
+    'java.lang.Integer',
+    'java.math.BigDecimal',
+    'java.util.HashMap',
+    'java.util.LinkedHashMap',
+    'java.util.LinkedList',
+]
+
+
+class JDBCBackend(Backend):
```

Comment:
I would not expose directly any references to JDBC essence. It should be hidden for consumer. E.g. in some future we may implement (java side) some additional storage options (nosql, file-based etc) which is not related to JDBC, but expose exactly same API (to python/R consumers).

Reply:
As mentioned in person, ixmp.Platform needs to choose and instantiate one Backend.
- For now, JDBCBackend is the only kind.
- Other Backends may provide storage without interacting with ixmp_source or any Java code.
- In the future, maybe, continued use of the "dbprops" and "dbtype" arguments could imply to use JDBCBackend, although I would prefer to clean up the semantics here.